### PR TITLE
implement loading ggml lora adapters through the modelfile

### PR DESCRIPTION
**Pull request by mxyng**
LoRA adapters can be added to Ollama models through the Modelfile and automatically applied when the model is loaded:

```
FROM llama2:13b
TEMPLATE {{ .Prompt }}
ADAPTER ./llama2-13b-storywriter-lora.ggml.bin
```

A few caveats:
* LoRA adapters must be GGML. If the adapter isn't GGML, it can be converted with the `convert-lora-to-ggml.py` script in https://github.com/ggerganov/llama.cpp
* Using adapters with quantized weights might not produce good results
* Using adapters disables mmap
* It's possible to apply multiple adapters but
  1. Ordering is important
  2. There may be unintended side effects since most adapters are not intended to be layered on other adapters
  3. Performance may degrade with more adapters

### CODE REVIEW

File: server/images.go [hunk 1]
```diff
@@ -294,6 +297,40 @@ func CreateModel(name string, path string, fn func(resp api.ProgressResponse)) e
 					layers = append(layers, newLayer)
 				}
 			}
+		case "adapter":
```
Comment (mxyng) [inline server/images.go:300]: A lot of this code is copied from the model code path. #288 refactors this into a separate function

File: llm/llama.go [hunk 2]
```diff
@@ -158,6 +158,12 @@ func newLlama(model string, opts api.Options) (*llama, error) {
 	params.use_mmap = C.bool(llm.UseMMap)
 	params.use_mlock = C.bool(llm.UseMLock)
 	params.embedding = C.bool(llm.EmbeddingOnly)
+
+	if len(adapters) > 0 {
+		log.Printf("must disable mmap to use lora adapters")
+		params.use_mmap = C.bool(false)
```
Comment (jmorganca) [inline llm/llama.go:164]: nice, this is a great way to handle this w/o full out erroring

File: server/images.go [hunk 3]
```diff
@@ -294,6 +297,40 @@ func CreateModel(name string, path string, fn func(resp api.ProgressResponse)) e
...
+				if err != nil {
+					return fmt.Errorf("failed to open file: %v", err)
+				}
+
+				fp = filepath.Join(home, filepath.Join(parts[1:]...))
+			}
+
+			// If filePath is not an absolute path, make it relative to the modelfile path
+			if !filepath.IsAbs(fp) {
+				fp = filepath.Join(filepath.Dir(path), fp)
```
Comment (jmorganca) [inline server/images.go:316]: 👍

**Review by jmorganca — APPROVED** [review]
This is a brilliant first step to supporting fine tuned layers.. nice!

**Review by pdevine — APPROVED** [review]
Really excited about this change!

File: docs/modelfile.md [hunk 4]
```diff
@@ -163,4 +172,4 @@ LICENSE """
 ## Notes
 
 - the **modelfile is not case sensitive**. In the examples, we use uppercase for instructions to make it easier to distinguish it from arguments.
-- Instructions can be in any order. In the examples, we start with FROM instruction to keep it easily readable.
\ No newline at end of file
+- Instructions can be in any order. In the examples, we start with FROM instruction to keep it easily readable.
```
[the remark this answers is not in the document]
Comment (mxyng) [reply docs/modelfile.md:175]: It's probably a newline that wasn't there before

**Pull request merged**